sidechain/consensus/common/src/peer_block_sync.rs

```diff
@@ -128,27 +133,34 @@ where
 	) -> Result<ParentchainBlock::Header> {
 		let shard_identifier = sidechain_block.block().header().shard_id();
 		let sidechain_block_number = sidechain_block.block().header().block_number();
+		let sidechain_block_hash = sidechain_block.hash();
 
 		// Attempt to import the block - in case we encounter an ancestry error, we go into
 		// peer fetching mode to fetch sidechain blocks from a peer and import those first.
-		match self.importer.import_block(sidechain_block, current_parentchain_header) {
+		match self.importer.import_block(sidechain_block.clone(), current_parentchain_header) {
 			Err(e) => match e {
 				Error::BlockAncestryMismatch(_block_number, block_hash, _) => {
 					warn!("Got ancestry mismatch error upon block import. Attempting to fetch missing blocks from peer");
-					self.fetch_and_import_blocks_from_peer(
+					let updated_parentchain_header = self.fetch_and_import_blocks_from_peer(
 						block_hash,
+						sidechain_block_hash,
 						current_parentchain_header,
 						shard_identifier,
-					)
+					)?;
+
+					self.importer.import_block(sidechain_block, &updated_parentchain_header)
```

Comment:
After importing the blocks fetched from a peer, we attempt another import of the original block. Before, the peer fetching imported all available blocks, which included the one we originally tried and failed - but that is no longer the case now.